core/src/main/java/com/linecorp/armeria/common/SerializationFormat.java

```diff
@@ -16,122 +16,227 @@
 
 package com.linecorp.armeria.common;
 
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.net.MediaType.create;
+import static java.util.Objects.requireNonNull;
 
+import java.util.Collection;
 import java.util.Collections;
-import java.util.EnumSet;
-import java.util.HashMap;
 import java.util.Locale;
-import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Optional;
+import java.util.ServiceLoader;
 import java.util.Set;
 
 import javax.annotation.Nullable;
 
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
 import com.google.common.net.MediaType;
 
 /**
  * Serialization format of a remote procedure call and its reply.
  */
-public enum SerializationFormat {
-    // FIXME(trustin): Make serialization formats registered via SPI instead of using non-extensible enum.
+public final class SerializationFormat implements Comparable<SerializationFormat> {
+
+    private static final BiMap<String, SerializationFormat> uriTextToFormats = HashBiMap.create();
```

Comment:
Sure. I can't use Stream or ImmutableBiMap.Builder though because it has to do duplicate checks. Do you have any suggestions for getting this done without wrapping a mutable map?